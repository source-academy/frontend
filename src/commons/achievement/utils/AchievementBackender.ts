import {
  AchievementGoal,
  AchievementItem,
  GoalDefinition,
  GoalMeta,
  GoalType
} from '../../../features/achievement/AchievementTypes';

export const backendifyGoalDefinition = (goal: GoalDefinition) => ({
  targetCount: goal.meta.type === GoalType.ASSESSMENT ? 1 : goal.meta.targetCount,
  meta: goal.meta,
  text: goal.text,
  type: goal.meta.type,
  uuid: goal.uuid
});

export const frontendifyAchievementGoal = (goal: any) =>
  ({
    uuid: goal.uuid || '',
    text: goal.text || '',
    achievementUuids: goal.achievementUuids,
    meta: (goal.meta.type === 'Event'
      ? {
          ...goal.meta,
          release: goal.meta.release ? new Date(goal.meta.release) : undefined,
          deadline: goal.meta.deadline ? new Date(goal.meta.deadline) : undefined,
          observeFrom: goal.meta.observeFrom ? new Date(goal.meta.observeFrom) : undefined,
          observeTo: goal.meta.observeTo ? new Date(goal.meta.observeTo) : undefined
        }
      : goal.meta) as GoalMeta,
    count: goal.count,
    targetCount: goal.targetCount,
    completed: goal.count >= goal.targetCount
  } as AchievementGoal);

export const frontendifyAchievementItem = (achievement: any) =>
  ({
    uuid: achievement.uuid || '',
    title: achievement.title || '',
    xp: achievement.xp,
    isVariableXp: achievement.isVariableXp,
    deadline: achievement.deadline === null ? undefined : new Date(achievement.deadline),
    release: achievement.release === null ? undefined : new Date(achievement.release),
    isTask: achievement.isTask,
    position: achievement.position,
    prerequisiteUuids: achievement.prerequisiteUuids,
    goalUuids: achievement.goalUuids,
    cardBackground: achievement.cardBackground || '',
    view: {
      coverImage: achievement.view.coverImage || '',
      completionText: achievement.view.completionText || '',
      description: achievement.view.description || ''
    }
  } as AchievementItem);

export const backendifyAchievementItem = (achievement: AchievementItem) => ({
  ...achievement,
  deadline: achievement.deadline ? achievement.deadline : null,
  release: achievement.release ? achievement.release : null
});
