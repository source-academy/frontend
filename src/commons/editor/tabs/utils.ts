import path from 'path';

/**
 * Returns the shortest file paths that is uniquely identifiable among
 * all open editor tabs. This is similar to how most code editors available
 * handle the displaying of file names.
 *
 * For example, if there are 3 open editor tabs where the file names are
 * exactly the same, we would need to display more of the file path to be
 * able to distinguish which editor tab corresponds to which file. Given
 * the following absolute file paths:
 * - /a.js
 * - /dir1/a.js
 * - /dir1/dir2/a.js
 * The shortest unique file paths will be:
 * - /a.js
 * - /dir1/a.js
 * - /dir2/a.js
 *
 * @param originalFilePaths The file paths to be shortened.
 */
export const getShortestUniqueFilePaths = (originalFilePaths: string[]): string[] => {
  // Store the unique shortened file paths as a mapping from the original file paths
  // to the shortened file paths. This is necessary because the output of this function
  // must preserve the original ordering of file paths.
  const originalToUniqueShortenedFilePaths: Record<string, string> = {};
  // Split each original file path into path segments and store the mapping from file
  // path to path segments for O(1) lookup. Since we only deal with the BrowserFS file
  // system, the path separator will always be '/'.
  const filePathSegments: Record<string, string[]> = originalFilePaths.reduce(
    (segments, filePath) => ({
      ...segments,
      // It is necessary to remove empty segments to deal with the very first '/' in
      // file paths.
      [filePath]: filePath.split(path.sep).filter(segment => segment !== '')
    }),
    {}
  );

  for (
    let numOfPathSegments = 1;
    // Keep looping while some original file paths have yet to be shortened.
    Object.keys(originalToUniqueShortenedFilePaths).length < originalFilePaths.length;
    numOfPathSegments++
  ) {
    // Based on the number of path segments for the iteration, we construct the
    // shortened file path. We then store the mapping from the shortened file path
    // to any original file path which transforms into it.
    const shortenedToOriginalFilePaths: Record<string, string[]> = Object.entries(
      filePathSegments
    ).reduce((filePaths, [originalFilePath, filePathSegments]) => {
      // Note that if there are fewer path segments than the number being sliced,
      // all of the path segments will be returned without error.
      const shortenedFilePath = '/' + filePathSegments.slice(-numOfPathSegments).join('/');
      return {
        ...filePaths,
        [shortenedFilePath]: (filePaths[shortenedFilePath] ?? []).concat(originalFilePath)
      };
    }, {});
    // Each shortened file path that only has a single corresponding original file
    // path is added to the unique shortened file paths record and their entry in
    // the file path segments record is removed to prevent further processing.
    Object.entries(shortenedToOriginalFilePaths).forEach(
      ([shortenedFilePath, originalFilePaths]) => {
        if (originalFilePaths.length > 1) {
          return;
        }

        const originalFilePath = originalFilePaths[0];
        originalToUniqueShortenedFilePaths[originalFilePath] = shortenedFilePath;
        // Remove the file path's segments from the next iteration.
        delete filePathSegments[originalFilePath];
      }
    );
  }

  // Finally, we retrieve the unique shortened file paths while preserving the ordering
  // of file paths.
  return originalFilePaths.map(filePath => originalToUniqueShortenedFilePaths[filePath]);
};
