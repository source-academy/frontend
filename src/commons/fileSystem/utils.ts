import { FSModule } from 'browserfs/dist/node/core/FS';
import Stats from 'browserfs/dist/node/core/node_fs_stats';
import path from 'path';

import { WORKSPACE_BASE_PATHS } from '../../pages/fileSystem/createInBrowserFileSystem';
import { WorkspaceLocation } from '../workspace/WorkspaceTypes';

type File = {
  path: string;
  contents: string;
};

/**
 * Retrieves the files in the specified workspace as a record that maps from
 * file path to file content. Because BrowserFS lacks an equivalent to Node.js's
 * Promises API, we need to wrap each asynchronous call to the file system with
 * promises ourselves.
 *
 * @param workspaceLocation The location of the workspace.
 * @param fileSystem        The file system instance.
 */
export const retrieveFilesInWorkspaceAsRecord = (
  workspaceLocation: WorkspaceLocation,
  fileSystem: FSModule
): Promise<Record<string, string>> => {
  const processFile = (filePath: string): Promise<File> => {
    return new Promise((resolve, reject) => {
      fileSystem.readFile(filePath, 'utf-8', (err, fileContents) => {
        if (err) {
          reject(err);
          return;
        }
        if (fileContents === undefined) {
          return;
        }

        resolve({
          path: filePath,
          contents: fileContents
        });
      });
    });
  };

  const processDirectory = (directoryPath: string): Promise<File[]> => {
    return new Promise((resolve, reject) => {
      fileSystem.readdir(directoryPath, (err, fileNames) => {
        if (err) {
          reject(err);
          return;
        }
        if (fileNames === undefined) {
          return;
        }

        const promises: Array<Promise<File | File[]>> = [];
        for (const fileName of fileNames) {
          const fullPath = path.join(directoryPath, fileName);
          promises.push(
            new Promise((resolve, reject) => {
              fileSystem.lstat(fullPath, (err, stats) => {
                if (err) {
                  reject(err);
                  return;
                }
                if (stats === undefined) {
                  return;
                }

                if (stats.isFile()) {
                  resolve(processFile(fullPath));
                } else if (stats.isDirectory()) {
                  resolve(processDirectory(fullPath));
                }
              });
            })
          );
        }

        Promise.all(promises).then((values: Array<File | File[]>) => {
          resolve(values.flat());
        });
      });
    });
  };

  const files = processDirectory(WORKSPACE_BASE_PATHS[workspaceLocation]);
  // Convert from array to record.
  return files.then((files: File[]) => {
    return files.reduce((record: Record<string, string>, file: File) => {
      record[file.path] = file.contents;
      return record;
    }, {});
  });
};

/**
 * Overwrites the files in the workspace with the ones specified. Because BrowserFS
 * lacks an equivalent to Node.js's Promises API, we need to wrap each asynchronous
 * call to the file system with promises ourselves.
 *
 * @param workspaceLocation The location of the workspace.
 * @param fileSystem        The file system instance.
 * @param files             A mapping from file paths to file contents.
 */
export const overwriteFilesInWorkspace = (
  workspaceLocation: WorkspaceLocation,
  fileSystem: FSModule,
  files: Record<string, string>
): Promise<void> => {
  return rmFilesInDirRecursively(fileSystem, WORKSPACE_BASE_PATHS[workspaceLocation]).then(() => {
    return new Promise(async (resolve, reject) => {
      for (const [filePath, fileContents] of Object.entries(files)) {
        await writeFileRecursively(fileSystem, filePath, fileContents).catch(err => reject(err));
      }
      resolve();
    });
  });
};

/**
 * Removes the files & directories in a directory recursively, but leave
 * the directory itself intact.
 *
 * @param fileSystem    The file system instance.
 * @param directoryPath The path of the directory to be removed.
 */
export const rmFilesInDirRecursively = (
  fileSystem: FSModule,
  directoryPath: string
): Promise<void> => {
  return new Promise((resolve, reject) => {
    fileSystem.readdir(directoryPath, async (err, fileNames): Promise<void> => {
      if (err) {
        reject(err);
        return;
      }
      if (fileNames === undefined) {
        return;
      }

      const removeFile = (fileName: string): Promise<Stats> => {
        const fullPath = path.join(directoryPath, fileName);
        return new Promise((resolve, reject) => {
          fileSystem.lstat(fullPath, async (err, stats) => {
            if (err) {
              reject(err);
              return;
            }
            if (stats === undefined) {
              return;
            }

            if (stats.isFile()) {
              await new Promise((resolve, reject) => {
                fileSystem.unlink(fullPath, err => {
                  if (err) {
                    reject(err);
                    return;
                  }
                  resolve(true);
                });
              });
            } else if (stats.isDirectory()) {
              await rmdirRecursively(fileSystem, fullPath);
            }

            resolve(stats);
          });
        });
      };

      // Remove each file/directory in parallel.
      Promise.all(fileNames.map(removeFile))
        .then(() => resolve())
        .catch(err => reject(err));
    });
  });
};

/**
 * Removes a directory recursively.
 *
 * BrowserFS's `rmdir` function is unable to remove non-empty directories despite being modelled
 * after Node.js's file system API. In Node.js, the `rmdir` function takes in an `options` object
 * where recursive directory removal can be set (https://nodejs.org/api/fs.html#fspromisesrmdirpath-options).
 * The BrowserFS equivalent however does not support these options and will fail with ENOTEMPTY
 * when trying to remove a non-empty directory.
 *
 * @param fileSystem    The file system instance.
 * @param directoryPath The path of the directory to be removed.
 */
export const rmdirRecursively = (fileSystem: FSModule, directoryPath: string): Promise<void> => {
  return new Promise((resolve, reject) => {
    rmFilesInDirRecursively(fileSystem, directoryPath)
      .then(() => {
        fileSystem.rmdir(directoryPath, err => {
          if (err) {
            reject(err);
            return;
          }
          resolve();
        });
      })
      .catch(err => reject(err));
  });
};

/**
 * Writes a file recursively by creating all parent directories if they do not exist.
 *
 * BrowserFS's `mkdir` function is unable to recursively create directories despite being modelled
 * after Node.js's file system API. In Node.js, the `mkdir` function takes in an `options` object
 * where recursive directory creation can be set (https://nodejs.org/api/fs.html#fspromisesmkdirpath-options).
 * The BrowserFS equivalent however does not support these options and will fail with ENOENT
 * when trying to create directories recursively.
 *
 * @param fileSystem   The file system instance.
 * @param filePath     The path of the file to be written to.
 * @param fileContents The contents of the file to be written to.
 */
export const writeFileRecursively = (
  fileSystem: FSModule,
  filePath: string,
  fileContents: string
): Promise<void> => {
  return new Promise((resolve, reject) => {
    // Create directories along the path if they do not exist.
    // Remove empty path segments.
    const pathSegments = filePath.split(path.sep).filter(path => path !== '');
    // The last path segment is not a directory but the name of the file, so we ignore it.
    for (let i = 1; i < pathSegments.length; i++) {
      const dirPath = '/' + path.join(...pathSegments.slice(0, i));
      const promise: Promise<void> = new Promise((resolve, reject) => {
        // Check whether the directory path already exists to prevent overwriting of existing files & directories.
        fileSystem.exists(dirPath, dirPathExists => {
          if (dirPathExists) {
            resolve();
            return;
          }

          fileSystem.mkdir(dirPath, 777, err => {
            if (err) {
              reject(err);
              return;
            }

            resolve();
          });
        });
      });
      promise.catch(err => reject(err));
    }

    // Write to the file.
    fileSystem.writeFile(filePath, fileContents, err => {
      if (err) {
        reject(err);
        return;
      }

      resolve();
    });
  });
};
