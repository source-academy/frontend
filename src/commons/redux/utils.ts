import {
  ActionCreatorWithOptionalPayload,
  ActionCreatorWithoutPayload,
  ActionCreatorWithPreparedPayload,
  createAction
} from '@reduxjs/toolkit';
import * as Sentry from '@sentry/browser';
import { SagaIterator } from 'redux-saga';
import { StrictEffect, takeEvery } from 'redux-saga/effects';

/**
 * Creates actions, given a base name and base actions
 * @param baseName The base name of the actions
 * @param baseActions The base actions. Use a falsy value to create an action without a payload.
 * @returns An object containing the actions
 */
export function createActions<BaseName extends string, BaseActions extends Record<string, any>>(
  baseName: BaseName,
  baseActions: BaseActions
) {
  return Object.entries(baseActions).reduce(
    (res, [name, func]) => ({
      ...res,
      [name]: func
        ? createAction(`${baseName}/${name}`, (...args: any) => ({ payload: func(...args) }))
        : createAction(`${baseName}/${name}`)
    }),
    {} as {
      [K in keyof BaseActions]: K extends string
        ? BaseActions[K] extends (...args: any) => any
          ? ActionCreatorWithPreparedPayload<
              Parameters<BaseActions[K]>,
              ReturnType<BaseActions[K]>,
              `${BaseName}/${K}`
            >
          : ActionCreatorWithoutPayload<`${BaseName}/${K}`>
        : never;
    }
  );
}

export function combineSagaHandlers<
  TActions extends Record<
    string,
    ActionCreatorWithPreparedPayload<any, any> | ActionCreatorWithoutPayload<any>
  >
>(
  actions: TActions,
  handlers: {
    // TODO: Maybe this can be stricter? And remove the optional type after
    // migration is fully done
    [K in keyof TActions]?: (action: ReturnType<TActions[K]>) => SagaIterator;
  },
  others?: (takeEvery: typeof saferTakeEvery) => SagaIterator
): () => SagaIterator {
  const sagaHandlers = Object.entries(handlers).map(([actionName, saga]) =>
    saferTakeEvery(actions[actionName], saga)
  );
  return function* (): SagaIterator {
    yield* sagaHandlers;
    if (others) {
      const obj = others(saferTakeEvery);
      while (true) {
        const { done, value } = obj.next();
        if (done) break;
        yield value;
      }
    }
  };
}

export function saferTakeEvery<
  Action extends
    | ActionCreatorWithOptionalPayload<any>
    | ActionCreatorWithPreparedPayload<any[], any>
>(actionPattern: Action, fn: (action: ReturnType<Action>) => Generator<StrictEffect<any>>) {
  function* wrapper(action: ReturnType<Action>) {
    try {
      yield* fn(action);
    } catch (error) {
      handleUncaughtError(error);
    }
  }

  return takeEvery(actionPattern.type, wrapper);
}

function handleUncaughtError(error: any) {
  if (process.env.NODE_ENV === 'development') {
    // react-error-overlay is a "special" package that's automatically included
    // in development mode by CRA

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    import('react-error-overlay').then(reo => reo.reportRuntimeError(error));
  }
  Sentry.captureException(error);
  console.error(error);
}
