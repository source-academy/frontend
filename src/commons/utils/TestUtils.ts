import React from 'react';
import renderer from 'react-test-renderer';
import { createRenderer } from 'react-test-renderer/shallow';

export const shallowRender = (element: React.ReactElement) => {
  const renderer = createRenderer();
  renderer.render(element);
  return renderer.getRenderOutput();
};

export const renderTreeJson = (element: React.ReactElement) => {
  return renderer.create(element).toJSON();
};

export const renderTree = (element: React.ReactElement) => {
  return renderer.create(element);
};

/**
 * Recursively traverses a nested object and returns all matches.
 *
 * Used in traversing nested objects such as `shallowRender`'s output.
 */
export function deepFilter<T>(
  nestedObject: T,
  matchFn: (e: T) => boolean,
  getChildren: (e: T) => T[]
) {
  const matches: any[] = [];

  function helper(obj: T) {
    if (matchFn(obj)) {
      matches.push(obj);
    }

    const children = getChildren(obj);
    if (children && Array.isArray(children)) {
      children.forEach(e => {
        if (e) {
          helper(e);
        }
      });
    }
  }

  helper(nestedObject);
  return matches;
}
