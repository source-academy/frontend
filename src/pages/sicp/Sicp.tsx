import 'katex/dist/katex.min.css';

import { Button, Classes, NonIdealState, Spinner } from '@blueprintjs/core';
import classNames from 'classnames';
import * as React from 'react';
import { useDispatch } from 'react-redux';
import { RouteComponentProps, useHistory, useParams } from 'react-router';
import Constants from 'src/commons/utils/Constants';
import { setLocalStorage } from 'src/commons/utils/LocalStorageHelper';
import { resetWorkspace, toggleUsingSubst } from 'src/commons/workspace/WorkspaceActions';
import { parseArr, ParseJsonError } from 'src/features/sicp/parser/ParseJson';
import { getNext, getPrev } from 'src/features/sicp/TableOfContentsHelper';
import {
  readSicpSectionLocalStorage,
  setSicpSectionLocalStorage,
  SICP_CACHE_KEY,
  SICP_INDEX
} from 'src/features/sicp/utils/SicpUtils';

import SicpErrorBoundary from '../../features/sicp/errors/SicpErrorBoundary';
import getSicpError, { SicpErrorType } from '../../features/sicp/errors/SicpErrors';
import SicpIndexPage from './subcomponents/SicpIndexPage';

type SicpProps = RouteComponentProps<{}>;

const baseUrl = Constants.sicpBackendUrl + 'json/';
const extension = '.json';

// Context to determine which code snippet is active
export const CodeSnippetContext = React.createContext({
  active: '0',
  setActive: (x: string) => {}
});

const loadingComponent = <NonIdealState title="Loading Content" icon={<Spinner />} />;

const Sicp: React.FC<SicpProps> = props => {
  const [data, setData] = React.useState(<></>);
  const [loading, setLoading] = React.useState(false);
  const [active, setActive] = React.useState('0');
  const { section } = useParams<{ section: string }>();
  const parentRef = React.useRef<HTMLDivElement>(null);
  const refs = React.useRef({});
  const history = useHistory();

  const scrollRefIntoView = (ref: HTMLDivElement | null) => {
    if (!ref || !parentRef?.current) {
      return;
    }

    const parent = parentRef.current!;
    const relativeTop = window.scrollY > parent.offsetTop ? window.scrollY : parent.offsetTop;

    parent.scrollTo({
      behavior: 'smooth',
      top: ref.offsetTop - relativeTop
    });
  };

  // Handle loading of latest viewed section and fetch json data
  React.useEffect(() => {
    if (!section) {
      /**
       * Handles rerouting to the latest viewed section when clicking from
       * the main application navbar. history.replace is used to allow the
       * user to still use the browser back button to navigate the app.
       */
      history.replace(`/sicpjs/${readSicpSectionLocalStorage()}`);
      return;
    }

    if (section === SICP_INDEX) {
      setSicpSectionLocalStorage(SICP_INDEX);
      return;
    }

    setLoading(true);

    fetch(baseUrl + section + extension)
      .then(response => {
        if (!response.ok) {
          throw Error(response.statusText);
        }
        return response.json();
      })
      .then(myJson => {
        try {
          const newData = parseArr(myJson, refs); // Might throw error
          setData(newData);
          setSicpSectionLocalStorage(section); // Sets local storage if valid page
        } catch (error) {
          throw new ParseJsonError(error.message);
        }
      })
      .catch(error => {
        console.error(error);

        if (error.message === 'Not Found') {
          // page not found
          setData(getSicpError(SicpErrorType.PAGE_NOT_FOUND_ERROR));
        } else if (error instanceof ParseJsonError) {
          // error occurred while parsing JSON
          setData(getSicpError(SicpErrorType.PARSING_ERROR));
        } else {
          setData(getSicpError(SicpErrorType.UNEXPECTED_ERROR));
        }
        setLocalStorage(SICP_CACHE_KEY, SICP_INDEX); // Prevents caching invalid page
      })
      .finally(() => {
        setLoading(false);
      });
  }, [section, history]);

  // Scroll to correct position
  React.useEffect(() => {
    if (loading) {
      return;
    }

    const hash = props.location.hash;
    const ref = refs.current[hash];

    scrollRefIntoView(ref);
  }, [props.location.hash, loading]);

  // Close all active code snippet when new page is loaded
  React.useEffect(() => {
    setActive('0');
  }, [data]);

  const dispatch = useDispatch();
  const handleSnippetEditorOpen = (s: string) => {
    setActive(s);
    dispatch(resetWorkspace('sicp'));
    dispatch(toggleUsingSubst(false, 'sicp'));
  };
  const handleNavigation = (sect: string) => {
    history.push('/sicpjs/' + sect);
  };

  const navigationButtons = (
    <div className="sicp-navigation-buttons">
      {getPrev(section) && (
        <Button onClick={() => handleNavigation(getPrev(section)!)}>Previous</Button>
      )}
      {getNext(section) && (
        <Button onClick={() => handleNavigation(getNext(section)!)}>Next</Button>
      )}
    </div>
  );

  return (
    <div
      className={classNames('Sicp', Classes.RUNNING_TEXT, Classes.TEXT_LARGE, Classes.DARK)}
      ref={parentRef}
    >
      <SicpErrorBoundary>
        <CodeSnippetContext.Provider value={{ active: active, setActive: handleSnippetEditorOpen }}>
          {loading ? (
            <div className="sicp-content">{loadingComponent}</div>
          ) : section === 'index' ? (
            <SicpIndexPage />
          ) : (
            <div className="sicp-content">
              {data}
              {navigationButtons}
            </div>
          )}
        </CodeSnippetContext.Provider>
      </SicpErrorBoundary>
    </div>
  );
};

export default Sicp;
