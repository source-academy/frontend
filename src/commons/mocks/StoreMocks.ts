import _ from 'lodash';
import { DeepPartial, Store } from 'redux';
import mockStore from 'redux-mock-store';

import {
  defaultAcademy,
  defaultAchievement,
  defaultApplication,
  defaultDashboard,
  defaultFileSystem,
  defaultPlayground,
  defaultSession,
  defaultWorkspaceManager,
  OverallState
} from '../application/ApplicationTypes';

export function mockInitialStore(overrides?: DeepPartial<OverallState>): Store<OverallState> {
  const createStore = (mockStore as any)();
  const state: OverallState = {
    academy: defaultAcademy,
    achievement: defaultAchievement,
    application: defaultApplication,
    dashboard: defaultDashboard,
    playground: defaultPlayground,
    workspaces: defaultWorkspaceManager,
    session: defaultSession,
    fileSystem: defaultFileSystem
  };
  return createStore(_.merge(state, overrides));
}
