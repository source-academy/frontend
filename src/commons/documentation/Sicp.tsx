import 'katex/dist/katex.min.css';

import { Button, Classes, NonIdealState, Spinner, TreeNodeInfo } from '@blueprintjs/core';
import classNames from 'classnames';
import React, { useRef, useState } from 'react';
import { useDispatch } from 'react-redux';
import { Link, useNavigate } from 'react-router';
import Constants from 'src/commons/utils/Constants';
import { setLocalStorage } from 'src/commons/utils/LocalStorageHelper';
import WorkspaceActions from 'src/commons/workspace/WorkspaceActions';
import { parseArr, ParseJsonError } from 'src/features/sicp/parser/ParseJson';
import { getNext, getPrev } from 'src/features/sicp/TableOfContentsHelper';
import {
  setSicpSectionLocalStorage,
  SICP_CACHE_KEY,
  SICP_INDEX
} from 'src/features/sicp/utils/SicpUtils';

import SicpErrorBoundary from '../../features/sicp/errors/SicpErrorBoundary';
import getSicpError, { SicpErrorType } from '../../features/sicp/errors/SicpErrors';
import SicpIndexPage from './SicpIndexPage';
import SicpNavigationBar from './SicpNavigationBar';

const baseUrl = Constants.sicpBackendUrl + 'json/';
const extension = '.json';

// Context to determine which code snippet is active
export const CodeSnippetContext = React.createContext({
  active: '0',
  setActive: (x: string) => {}
});

const loadingComponent = <NonIdealState title="Loading Content" icon={<Spinner />} />;

type SicpProps = {
  setSicpHomeCallBackFn: (fn: () => void) => void;
};

const Sicp: React.FC<SicpProps> = props => {
  const [data, setData] = useState(<></>);
  const [loading, setLoading] = useState(false);
  const [active, setActive] = useState('0');
  const [section, setSection] = useState('index');
  const [hash, setHash] = useState('#begin');
  const refs = useRef<Record<string, HTMLElement | null>>({});
  const navigate = useNavigate();

  const scrollRefIntoView = (ref: HTMLElement | null) => {
    // console.log(`Scrolling ${ref} into view...`);
    if (!ref) {
      return;
    }

    ref.scrollIntoView({
      behavior: 'smooth'
    });
  };

  // Handle loading of latest viewed section and fetch json data
  React.useEffect(() => {
    props.setSicpHomeCallBackFn(() => {
      setSection('index');
    });
    if (!section) {
      /**
       * Handles rerouting to the latest viewed section when clicking from
       * the main application navbar. Navigate replace logic is used to allow the
       * user to still use the browser back button to navigate the app.
       */
      // navigate(`/sicpjs/${readSicpSectionLocalStorage()}`, { replace: true });
      return;
    }

    if (section === SICP_INDEX) {
      setSicpSectionLocalStorage(SICP_INDEX);
      return;
    }

    setLoading(true);

    fetch(baseUrl + section + extension)
      .then(response => {
        if (!response.ok) {
          throw Error(response.statusText);
        }
        return response.json();
      })
      .then(myJson => {
        try {
          const newData = parseArr(myJson, refs); // Might throw error
          setData(newData);
          setSicpSectionLocalStorage(section); // Sets local storage if valid page
        } catch (error) {
          throw new ParseJsonError(error.message);
        }
      })
      .catch(error => {
        console.error(error);

        if (error.message === 'Not Found') {
          // page not found
          setData(getSicpError(SicpErrorType.PAGE_NOT_FOUND_ERROR));
        } else if (error instanceof ParseJsonError) {
          // error occurred while parsing JSON
          setData(getSicpError(SicpErrorType.PARSING_ERROR));
        } else {
          setData(getSicpError(SicpErrorType.UNEXPECTED_ERROR));
        }
        setLocalStorage(SICP_CACHE_KEY, SICP_INDEX); // Prevents caching invalid page
      })
      .finally(() => {
        setLoading(false);
      });
  }, [section, navigate, props]);

  // Scroll to correct position
  React.useEffect(() => {
    if (loading) {
      return;
    }

    const ref = refs.current[`#${hash}`];
    scrollRefIntoView(ref);
  }, [loading, hash]);

  // Close all active code snippet when new page is loaded
  React.useEffect(() => {
    setActive('0');
  }, [data]);

  const dispatch = useDispatch();
  const handleSnippetEditorOpen = (s: string) => {
    setActive(s);
    dispatch(WorkspaceActions.resetWorkspace('sicp'));
    dispatch(WorkspaceActions.toggleUsingSubst(false, 'sicp'));
  };
  const handleNavigation = (sect: string) => {
    // console.log(`Navigating to section: ${sect}`);
    setSection(sect);
  };

  // `section` is defined due to the navigate logic in the useEffect above
  const navigationButtons = (
    <div className="sicp-navigation-buttons" style={{ margin: '50px 0' }}>
      {getPrev(section!) && (
        <Button style={{ textWrap: 'nowrap' }} onClick={() => handleNavigation(getPrev(section!)!)}>
          Previous
        </Button>
      )}
      {getNext(section!) && (
        <Button style={{ textWrap: 'nowrap' }} onClick={() => handleNavigation(getNext(section!)!)}>
          Next
        </Button>
      )}
    </div>
  );

  const handleNodeClicked = React.useCallback(
    (node: TreeNodeInfo) => {
      // console.log(`Navigating to section: ${String(node.nodeData)}`);
      setSection(String(node.nodeData));
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [section]
  );

  const handleNodeClickedString = React.useCallback(
    (sect: string) => {
      const urlPart = sect.split('#');
      if (urlPart.length > 1) {
        console.log(`URL Part: ${urlPart[1]}`);
        setHash(urlPart[1]);
      }
      // console.log(`Navigating to section: ${urlPart[0]}`);
      setSection(urlPart[0]);
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [section]
  );

  return (
    <div style={{ height: '100%' }} className={classNames('Sicp', Classes.TEXT_LARGE)}>
      <SicpNavigationBar
        handleNodeClickedString={handleNodeClickedString}
        handleNodeClicked={handleNodeClicked}
      />
      <SicpErrorBoundary>
        <CodeSnippetContext.Provider value={{ active: active, setActive: handleSnippetEditorOpen }}>
          {loading ? (
            <div className="sicp-content">{loadingComponent}</div>
          ) : section === 'index' ? (
            <SicpIndexPage handleNodeClick={handleNodeClicked} />
          ) : (
            <div className="sicp-content sicp-assessment">
              <Link to="#begin" ref={ref => (refs.current['#begin'] = ref)} />
              {data}
              {navigationButtons}
              <Link to="#end" ref={ref => (refs.current['#end'] = ref)} />
            </div>
          )}
        </CodeSnippetContext.Provider>
      </SicpErrorBoundary>
    </div>
  );
};

// react-router lazy loading
// https://reactrouter.com/en/main/route/lazy
export const Component = Sicp;
Component.displayName = 'Sicp';

export default Sicp;
