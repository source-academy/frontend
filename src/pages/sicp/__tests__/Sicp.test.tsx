import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import type { Location } from 'react-router';
import { mockInitialStore } from 'src/commons/mocks/StoreMocks';
import { shallowRender } from 'src/commons/utils/TestUtils';
import { vi } from 'vitest';

import Sicp from '../Sicp';

vi.mock('react-router', () => ({
  useParams: vi.fn().mockReturnValue({ section: 'index' }),
  useNavigate: vi.fn().mockReturnValue(vi.fn()),
  useLocation: vi.fn().mockReturnValue({} as Location)
}));

describe('Sicp renders', () => {
  test('correctly', () => {
    const sicp = (
      <Provider store={mockInitialStore()}>
        <Sicp />
      </Provider>
    );
    const tree = shallowRender(sicp);
    expect(tree).toMatchSnapshot();
  });

  test('index section correctly', () => {
    const sicp = (
      <Provider store={mockInitialStore()}>
        <Sicp />
      </Provider>
    );
    const { container } = render(sicp);
    expect(container.querySelector('.sicp-index-page')).toBeTruthy();
  });
});
