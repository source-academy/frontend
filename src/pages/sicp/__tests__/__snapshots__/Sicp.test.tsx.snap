// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sicp renders correctly 1`] = `
<Context.Provider
  value={
    {
      "getServerState": undefined,
      "noopCheck": "once",
      "stabilityCheck": "once",
      "store": {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <Sicp />
</Context.Provider>
`;
